File: 프로그래머스/2/87946. 피로도/피로도.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

void dfs(int k, int count, int* answer, bool* visited, int** dungeons, int n) {
    if (count > *answer) {
        *answer = count;
    }
    
    for (int i = 0; i < n; i++) {
        if (!visited[i] && k >= dungeons[i][0]) {
            visited[i] = true;
            dfs(k - dungeons[i][1], count + 1, answer, visited, dungeons, n);
            visited[i] = false;
        }
    }
}

int solution(int k, int** dungeons, size_t dungeons_rows, size_t dungeons_cols) {
    int answer = 0;
    int n = dungeons_rows;
    
    bool* visited = (bool*)calloc(n, sizeof(bool));
    
    dfs(k, 0, &answer, visited, dungeons, n);
    
    free(visited);
    
    return answer;
}

File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    max_val = -1
    for p in permutations(dungeons):
        cost = k
        count = 0
        for i in p:
            if i[0] <= cost:
                count += 1
                cost -= i[1]
            else:
                break
        
        max_val = max(max_val, count)
    
    return max_val